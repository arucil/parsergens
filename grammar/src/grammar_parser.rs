
pub mod lex;
pub mod ast;

use lalrpop_util::{lalrpop_mod, ParseError};

lalrpop_mod!(pub grammar, "/grammar_parser/grammar.rs");

pub fn parse(
  input: &str
) -> Result<ast::Grammar, ParseError<usize, lex::Token, lex::LexError>> {
  let lexer = lex::Lexer::new(input);
  grammar::DocumentParser::new().parse(input, lexer)
}
