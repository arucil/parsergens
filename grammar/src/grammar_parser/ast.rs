use std::ops::Range;

pub type Grammar = Vec<Spanned<Decl>>;

#[derive(Debug)]
pub enum Decl {
  Token(TokenDecl),
  Start(StartDecl),
  Rule(RuleDecl),
  Skip(SkipDecl),
}

#[derive(Debug)]
pub struct TokenDecl {
  pub name: Spanned<String>,
  pub pattern: TokenPattern,
}

#[derive(Debug)]
pub enum TokenPattern {
  Regex(Spanned<(String, Regex)>),
  String(Spanned<(String, Regex)>),
}

#[derive(Debug)]
pub enum Regex {
  Any,
  Char(char),
  CharSet(Vec<Range<char>>),
  CharClass(CharClassKind),
  Or(Box<Regex>, Box<Regex>),
  Optional(Box<Regex>),
  Many(Box<Regex>),
  Many1(Box<Regex>),
  Seq(Vec<Regex>),
}

#[derive(Debug)]
pub enum CharClassKind {
  Digit,
  Word,
}

#[derive(Debug)]
pub struct SkipDecl {
  pub pattern: TokenPattern,
}

#[derive(Debug)]
pub struct StartDecl {
  pub name: Spanned<String>,
}

#[derive(Debug)]
pub struct RuleDecl {
  pub name: Spanned<String>,
  pub alts: Vec<Spanned<RuleAlt>>,
}

#[derive(Debug)]
pub enum RuleAlt {
  Epsilon,
  Terms(Vec<Spanned<String>>),
}

#[derive(Debug)]
pub struct Spanned<T>(pub Range<usize>, pub T);