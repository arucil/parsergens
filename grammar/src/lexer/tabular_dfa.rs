//! DFA table compression schem from
//! https://stackoverflow.com/questions/29139162/dfa-state-transition-table-compression

use bitvec::prelude::*;
use super::dfa::Dfa;
use crate::Map;

pub use super::dfa::State;

#[derive(Debug)]
pub struct TabularDfa<V> {
  start: u32,
  state_base: Vec<usize>,
  transitions: Vec<(u32, u32)>,
  accept_states: Map<State, V>,
}

impl<V> TabularDfa<V> {
  pub fn start(&self) -> State {
    State(self.start)
  }

  pub fn transition(&self, state: State, c: u32) -> Option<State> {
    let tx = self.transitions[self.state_base[state.0 as usize] + c as usize];
    if tx.0 == state.0 && tx.1 != 0 {
      Some(State(tx.1 - 1))
    } else {
      None
    }
  }

  /// if the state is accepting state, return the value.
  pub fn result(&self, state: State) -> Option<&V> {
    self.accept_states.get(&state)
  }
}

impl<V> From<Dfa<u32, V>> for TabularDfa<V> {
  fn from(dfa: Dfa<u32, V>) -> Self {
    let last_letter = dfa.transitions.keys()
      .map(|(_, c)| *c).max().unwrap();
    let last_state = dfa.transitions.iter()
      .map(|((s1, _), s2)| s1.0.max(s2.0))
      .max().unwrap()
      + 1;

    let mut state_base = vec![];
    let mut transitions = vec![];
    let mut row_mask = bitvec![0; last_letter as usize + 1];
    let mut base = 0;

    for state in 0..=last_state {
      let mut row_bits = BitVec::<LocalBits, usize>::with_capacity(
        (last_letter + 1) as usize);
      let mut row = vec![];

      for c in 0..=last_letter {
        if let Some(&State(next)) = dfa.transitions.get(&(State(state), c)) {
          row.push(next + 1);
          row_bits.push(true);
        } else {
          row.push(0);
          row_bits.push(false);
        }
      }

      loop {
        let overlap = row_mask.clone() & row_bits.clone();
        if overlap.not_any() {
          row_mask |= row_bits;
          break;
        }

        base += 1;
        row_mask.set(0, false);
        row_mask.rotate_left(1);
      }

      state_base.push(base);

      transitions.resize(base + last_letter as usize + 1, (state, 0));

      for (i, x) in row.into_iter().enumerate() {
        if x != 0 {
          transitions[base + i] = (state, x);
        }
      }

      while row_mask[0] {
        base += 1;
        row_mask.set(0, false);
        row_mask.rotate_left(1);
      }
    }

    Self {
      start: dfa.start,
      state_base,
      transitions,
      accept_states: dfa.accept_states,
    }
  }
}