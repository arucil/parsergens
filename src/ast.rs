use std::ops::Range;

pub type Grammar = Vec<Spanned<Decl>>;

#[derive(Debug)]
pub enum Decl {
  Token(TokenDecl),
  Start(StartDecl),
  Rule(RuleDecl),
}

#[derive(Debug)]
pub struct TokenDecl {
  pub name: Spanned<String>,
  pub regex: Spanned<String>,
}

#[derive(Debug)]
pub struct StartDecl {
  pub name: Spanned<String>,
}

#[derive(Debug)]
pub struct RuleDecl {
  pub name: Spanned<String>,
  pub alts: Vec<Spanned<RuleAlt>>,
}

#[derive(Debug)]
pub enum RuleAlt {
  Epsilon,
  Terms(Vec<Spanned<String>>),
}

#[derive(Debug)]
pub struct Spanned<T>(pub Range<usize>, pub T);