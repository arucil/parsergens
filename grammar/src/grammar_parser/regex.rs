use std::str::CharIndices;
use std::iter::Peekable;
use either::Either::{self, *};
use super::ast::{Regex, CharClass, CharSetItem};

#[derive(Debug)]
pub struct RegexError {
  pub kind: RegexErrorKind,
  pub span: (usize, usize),
}

#[derive(Debug)]
pub enum RegexErrorKind {
  SyntaxError,
  InvalidCodePoint,
  Empty,
}

type Result<T=Regex> = std::result::Result<T, RegexError>;

pub fn parse_string_literal(
  input: &str,
  start: usize,
) -> Result {
  RegexParser::new(input)
    .parse_string_literal()
    .map_err(|err| {
      RegexError {
        kind: err.kind,
        span: (err.span.0 + start, err.span.1 + start),
      }
    })
}

pub fn parse_regex(
  input: &str,
  start: usize,
) -> Result {
  RegexParser::new(input)
    .parse()
    .map_err(|err| {
      RegexError {
        kind: err.kind,
        span: (err.span.0 + start, err.span.1 + start),
      }
    })
}

struct RegexParser<'a> {
  input: &'a str,
  chars: Peekable<CharIndices<'a>>,
}

impl<'a> RegexParser<'a> {
  fn new(input: &'a str) -> Self {
    Self {
      input,
      chars: input.char_indices().peekable()
    }
  }

  fn parse(&mut self) -> Result {
    let regex = self.parse_regex()?;

    if let Some((i, _)) = self.chars.next() {
      syntax_error(i, self.input.len())?
    } else {
      Ok(regex)
    }
  }

  fn parse_string_literal(&mut self) -> Result {
    let mut items = vec![];

    while let Some((i0, c)) = self.chars.next() {
      if c == '\\' {
        match self.parse_escape(i0, false)? {
          Left(c) => items.push(Regex::Char(c)),
          _ => unreachable!(),
        }
      } else {
        items.push(Regex::Char(c));
      }
    }

    if items.is_empty() {
      Ok(Regex::Empty)
    } else if items.len() == 1 {
      Ok(items.pop().unwrap())
    } else {
      Ok(Regex::Concat(items))
    }
  }

  fn parse_regex(&mut self) -> Result {
    self.parse_alt()
  }

  fn parse_alt(&mut self) -> Result {
    let mut items = vec![self.parse_concat()?];

    while let Some((_, '|')) = self.chars.peek() {
      self.advance();
      items.push(self.parse_concat()?);
    }

    if items.len() == 1 {
      Ok(items.pop().unwrap())
    } else {
      Ok(Regex::Alt(items))
    }
  }

  fn parse_concat(&mut self) -> Result {
    let mut items = vec![self.parse_repetition()?];

    loop {
      if let Some((_, '|' | ')')) | None = self.chars.peek() {
        break;
      }

      items.push(self.parse_repetition()?);
    }

    if items.len() == 1 {
      Ok(items.pop().unwrap())
    } else {
      Ok(Regex::Concat(items))
    }
  }

  fn parse_repetition(&mut self) -> Result {
    let mut item = self.parse_factor()?;

    while let Some(&(_, c@('+' | '*' | '?'))) = self.chars.peek() {
      self.advance();
      item = match c {
        '+' => Regex::Many1(Box::new(item)),
        '*' => Regex::Many(Box::new(item)),
        '?' => Regex::Optional(Box::new(item)),
        _ => unreachable!(),
      };
    }

    Ok(item)
  }

  fn parse_factor(&mut self) -> Result {
    match self.chars.peek() {
      Some((_, '.')) => {
        self.advance();
        Ok(Regex::Any)
      }
      Some((_, '(')) => {
        self.advance();
        let regex = self.parse_regex()?;
        self.match_char(')')?;
        Ok(regex)
      }
      Some(&(i, '\\')) => {
        self.advance();
        match self.parse_escape(i, true)? {
          Left(c) => Ok(Regex::Char(c)),
          Right(class) => Ok(Regex::CharClass(class)),
        }
      }
      Some(&(i, '[')) => {
        self.advance();
        self.parse_char_set(i)
      }
      Some(&(i, c@('+' | '*' | '?'))) => {
        syntax_error(i, i + c.len_utf8())?
      }
      Some((_, '|' | ')')) | None => {
        Ok(Regex::Empty)
      }
      Some(&(_, c)) => {
        self.advance();
        Ok(Regex::Char(c))
      }
    }
  }

  fn parse_char_set(&mut self, _start: usize) -> Result {
    let mut pending_range = false;
    let mut items = vec![];

    loop {
      let item = match self.chars.next() {
        Some((_, ']')) => {
          if pending_range {
            items.push(CharSetItem::Char('-'));
          }
          return Ok(Regex::CharSet(items));
        }
        Some((i, '\\')) => {
          match self.parse_escape(i, true)? {
            Left(c) => CharSetItem::Char(c),
            Right(class) => CharSetItem::CharClass(class),
          }
        }
        Some((_, '-')) => {
          if pending_range {
            CharSetItem::Char('-')
          } else if let Some(CharSetItem::Char(_)) = items.last() {
            pending_range = true;
            continue;
          } else {
            CharSetItem::Char('-')
          }
        }
        Some((_, c)) => {
          CharSetItem::Char(c)
        }
        None => {
          self.eof_error()?
        }
      };

      if pending_range {
        match &item {
          CharSetItem::Char(c) => {
            let c0 = match items.pop().unwrap() {
              CharSetItem::Char(c) => c,
              _ => unreachable!(),
            };
            items.push(CharSetItem::Range(c0, *c));
          }
          _ => {
            items.push(CharSetItem::Char('-'));
            items.push(item);
          }
        }
        pending_range = false;
      } else {
        items.push(item);
      }
    }
  }

  fn parse_escape(
    &mut self,
    start: usize,
    enable_class: bool
  ) -> Result<Either<char, CharClass>> {
    match self.chars.next() {
      Some((_, 'n')) => {
        Ok(Left('\n'))
      }
      Some((_, 'r')) => {
        Ok(Left('\r'))
      }
      Some((_, 't')) => {
        Ok(Left('\t'))
      }
      Some((_, '\\')) => {
        Ok(Left('\\'))
      }
      Some((_, 'd')) if enable_class => {
        Ok(Right(CharClass::Digit))
      }
      Some((_, 'w')) if enable_class => {
        Ok(Right(CharClass::Word))
      }
      Some((i, 'U')) => {
        if let Some((_, '{')) = self.chars.next() {
          ()
        } else {
          syntax_error(start, i + 1)?
        }

        let cp_start = i + 2;
        let cp_end;
        loop {
          match self.chars.next() {
            Some((i, '}')) => {
              cp_end = i;
              break;
            }
            None => {
              syntax_error(start, self.input.len())?
            }
            _ => {}
          }
        }

        if cp_start == cp_end {
          syntax_error(start, cp_end)?
        } else {
          let ch = u32::from_str_radix(&self.input[cp_start..cp_end], 16)
            .map_err(|_| ())
            .and_then(|cp| std::char::from_u32(cp).ok_or(()))
            .map_err(|_| RegexError {
              kind: RegexErrorKind::InvalidCodePoint,
              span: (cp_start, cp_end),
            })?;

          Ok(Left(ch))
        }
      }
      Some((_, c)) => {
        Ok(Left(c))
      }
      None => {
        self.eof_error()?
      }
    }
  }

  fn match_char(&mut self, c: char) -> Result<char> {
    self.match_char_if(|x| x == c)
  }

  fn match_char_if(&mut self, f: impl Fn(char) -> bool) -> Result<char> {
    if let Some((i, c)) = self.chars.next() {
      if f(c) {
        Ok(c)
      } else {
        syntax_error(i, i + c.len_utf8())?
      }
    } else {
      self.eof_error()?
    }
  }

  fn advance(&mut self) {
    self.chars.next();
  }

  fn eof_error(&self) -> Result<!> {
    Err(RegexError {
      kind: RegexErrorKind::SyntaxError,
      span: (self.input.len(), self.input.len())
    })
  }
}

fn syntax_error(start: usize, end: usize) -> Result<!> {
  Err(RegexError {
    kind: RegexErrorKind::SyntaxError,
    span: (start, end),
  })
}


#[cfg(test)]
mod tests {
  use super::*;
  use insta::assert_debug_snapshot;

  mod string_literal {
    use super::*;

    #[test]
    fn trivial() {
      assert_debug_snapshot!(parse_string_literal("a.-+&Ro1", 100));
    }

    #[test]
    fn empty() {
      assert_debug_snapshot!(parse_string_literal("", 100));
    }

    #[test]
    fn single_char() {
      assert_debug_snapshot!(parse_string_literal("+", 100));
    }

    #[test]
    fn escape() {
      assert_debug_snapshot!(parse_string_literal(r"ab\U{41}\r\nn\tt\d", 100));
    }
  }

  mod regex {
    use super::*;

    #[test]
    fn empty() {
      assert_debug_snapshot!(parse_regex(r"", 100));
    }

    #[test]
    fn trivial() {
      assert_debug_snapshot!(parse_regex("aRo1", 100));
    }

    #[test]
    fn escape() {
      assert_debug_snapshot!(parse_regex(r"ab\U{41}\r\nn\tt\.\+\d", 100));
    }

    #[test]
    fn alt_empty() {
      assert_debug_snapshot!(parse_regex(r"||ab", 100));
    }

    #[test]
    fn concat_precedes_alt() {
      assert_debug_snapshot!(parse_regex(r"A1\.|\d|ab", 100));
    }

    #[test]
    fn repetition_precedence() {
      assert_debug_snapshot!(parse_regex(r"ab+\||.*|ab?|", 100));
    }

    #[test]
    fn multi_repetition() {
      assert_debug_snapshot!(parse_regex(r"ab+?|\d+*", 100));
    }

    #[test]
    fn parenthesis() {
      assert_debug_snapshot!(parse_regex(r"(ab|\w\d+)+|\.", 100));
    }

    #[test]
    fn char_set_trivial() {
      assert_debug_snapshot!(parse_regex(r"[1234567890]+", 100));
    }

    #[test]
    fn char_set_range() {
      assert_debug_snapshot!(parse_regex(r"[0-9]+", 100));
    }

    #[test]
    fn char_set_multi_range() {
      assert_debug_snapshot!(parse_regex(r"[0-9a-zA-Z]+", 100));
    }

    #[test]
    fn char_set_mixed() {
      assert_debug_snapshot!(parse_regex(r"[0-9abcA-Z.]+", 100));
    }

    #[test]
    fn char_set_class() {
      assert_debug_snapshot!(parse_regex(r"[a-zA-Z_][\w_]*", 100));
    }

    #[test]
    fn char_set_escape() {
      assert_debug_snapshot!(parse_regex(r"([a-zA-\U{45}_] )", 100));
    }

    #[test]
    fn char_set_hyphen() {
      assert_debug_snapshot!(parse_regex(r"([-[-z ---_-] )", 100));
    }

    #[test]
    fn repetition_mixed() {
      assert_debug_snapshot!(parse_regex(r"\d+(\.\d*)?", 100));
    }
  }
}