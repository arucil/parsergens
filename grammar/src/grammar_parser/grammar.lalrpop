use lalrpop_util::ParseError;
use super::ast::*;
use super::lex;
use super::UserParseError;
use super::regex;

#[LALR]
grammar<'input>(input: &'input str);

pub Document: Grammar = (Separator <decl>)*
  => {
    <>.into_iter()
      .filter_map(|x| x)
      .collect()
  };

decl: Option<Spanned<Decl>> = {
  <l:@L> <decl:rule_decl> <r:@L> => Some(Spanned((l, r), Decl::Rule(decl))),
  <l:@L> <decl:start_decl> <r:@L> => Some(Spanned((l, r), Decl::Start(decl))),
  <l:@L> <decl:token_decl> <r:@L> => Some(Spanned((l, r), Decl::Token(decl))),
  <l:@L> <decl:ext_token_decl> <r:@L> => Some(Spanned((l, r), Decl::ExtToken(decl))),
  <l:@L> <decl:skip_decl> <r:@L> => Some(Spanned((l, r), Decl::Skip(decl))),
  <l:@L> <decl:user_decl> <r:@L> => Some(Spanned((l, r), Decl::User(decl))),
  <l:@L> <decl:assoc_decl> <r:@L> => Some(Spanned((l, r), Decl::Assoc(decl))),
  <l:@L> <decl:state_decl> <r:@L> => Some(Spanned((l, r), Decl::State(decl))),
};

rule_decl: RuleDecl =
  <name:ident> <ty:(Colon <code_block>)?> Assign <alt1:production> <mut alts:(Or <production>)*>
  =>? {
    alts.insert(0, alt1);
    if alts.iter().any(|alt| alt.1.action.is_some()) && ty.is_none() {
      Err(ParseError::User {
        error: UserParseError::MissingType(name.0),
      })
    } else {
      Ok(RuleDecl {
        name,
        ty,
        alts,
      })
    }
  }; 

production: Spanned<RuleAlt> = {
  <l:@L> <terms:term+> <prec:(Percent Prec <ident>)?> <action:code_block?> <r:@L> =>
    Spanned(
      (l, r),
      RuleAlt {
        terms: RuleAltTerms::Terms(terms),
        prec,
        action,
      }
    ),
  <l:@L> epsilon <action:code_block?> <r:@L> =>
    Spanned(
      (l, r),
      RuleAlt {
        terms: RuleAltTerms::Epsilon,
        prec: None,
        action,
      }
    ),
};

epsilon: () = LParen RParen;

term: Term = {
  ident => Term::Symbol(<>),
  LParen <terms:term+> RParen QuestionMark => Term::Optional(terms),
  LParen <terms:term+> RParen Asterisk => Term::Many(terms),
  LParen <terms:term+> RParen Plus => Term::Many1(terms),
}

start_decl: StartDecl = Percent Start <name:ident>
  => {
    StartDecl {
      name,
    }
  };

token_decl: TokenDecl = Percent Token <name:ident> <pattern:token_pattern>
  => {
    TokenDecl {
      name,
      pattern,
    }
  };

ext_token_decl: ExternalTokenDecl = Percent Token <name:ident>
  => {
    ExternalTokenDecl {
      name,
    }
  };

skip_decl: SkipDecl = Percent Skip <pattern:token_pattern>
  => {
    SkipDecl {
      pattern,
    }
  };

user_decl: UserDecl = Percent User <code:code_block>
  => {
    UserDecl {
      code,
    }
  };

assoc_decl: AssocDecl = Percent <assoc:assoc> <names:ident+>
  => {
    AssocDecl {
      names,
      assoc,
    }
  };

assoc: Spanned<Assoc> = {
  <l:@L> LeftAssoc <r:@L> => Spanned((l, r), Assoc::LeftAssoc),
  <l:@L> RightAssoc <r:@L> => Spanned((l, r), Assoc::RightAssoc),
  <l:@L> NonAssoc <r:@L> => Spanned((l, r), Assoc::NonAssoc),
}

state_decl: StateDecl = Percent State <state:code_block>
  => {
    StateDecl {
      state,
    }
  };

token_pattern: TokenPattern = {
  <regex:regex> =>? regex::parse_regex(
    &regex.1[1..regex.1.len() - 1],
    regex.0 .0 + 1
  )
    .map(|x| TokenPattern {
      kind: TokenPatternKind::Regex,
      source: regex,
      regex: x,
    })
    .map_err(|error| ParseError::User {
      error: UserParseError::RegexError(error),
    }),
  <str:string> =>? regex::parse_string_literal(
    &str.1[1..str.1.len() - 1], str.0 .0
  )
    .map(|x| TokenPattern {
      kind: TokenPatternKind::String,
      source: str,
      regex: x,
    })
    .map_err(|error| ParseError::User {
      error: UserParseError::RegexError(error)
    }),
}

spanned<T>: Spanned<String> = <l:@L> <x:T> <r:@L> => Spanned((l, r), x.text.to_owned());

ident = spanned<Ident>;
string = spanned<String>;
regex = spanned<Regex>;
code_block: Spanned<String> = <spanned<CodeBlock>> =>
  Spanned(<>.0, <>.1.trim_start_matches('{').trim_end_matches('}').trim().to_owned());

extern {
  type Location = usize;
  type Error = UserParseError;

  enum lex::Token<'input> {
    Start => lex::Token { kind: lex::TokenKind::Start, .. },
    Token => lex::Token { kind: lex::TokenKind::Token, .. },
    Skip => lex::Token { kind: lex::TokenKind::Skip, .. },
    User => lex::Token { kind: lex::TokenKind::User, .. },
    LeftAssoc => lex::Token { kind: lex::TokenKind::LeftAssoc, .. },
    RightAssoc => lex::Token { kind: lex::TokenKind::RightAssoc, .. },
    NonAssoc => lex::Token { kind: lex::TokenKind::NonAssoc, .. },
    Prec => lex::Token { kind: lex::TokenKind::Prec, .. },
    State => lex::Token { kind: lex::TokenKind::State, .. },

    Ident => lex::Token { kind: lex::TokenKind::Ident, .. },
    Regex => lex::Token { kind: lex::TokenKind::Regex, .. },
    String => lex::Token { kind: lex::TokenKind::String, .. },
    CodeBlock => lex::Token { kind: lex::TokenKind::CodeBlock, .. },

    Percent => lex::Token { kind: lex::TokenKind::Percent, .. },
    Assign => lex::Token { kind: lex::TokenKind::Assign, .. },
    Or => lex::Token { kind: lex::TokenKind::Or, .. },
    LParen => lex::Token { kind: lex::TokenKind::LParen, .. },
    RParen => lex::Token { kind: lex::TokenKind::RParen, .. },
    Comma => lex::Token { kind: lex::TokenKind::Comma, .. },
    Colon => lex::Token { kind: lex::TokenKind::Colon, .. },
    Asterisk => lex::Token { kind: lex::TokenKind::Asterisk, .. },
    QuestionMark => lex::Token { kind: lex::TokenKind::QuestionMark, .. },
    Plus => lex::Token { kind: lex::TokenKind::Plus, .. },
    Arrow => lex::Token { kind: lex::TokenKind::Arrow, .. },

    Enter => lex::Token { kind: lex::TokenKind::Enter, .. },
    Leave => lex::Token { kind: lex::TokenKind::Leave, .. },
    Separator => lex::Token { kind: lex::TokenKind::Separator, .. },
  }
}