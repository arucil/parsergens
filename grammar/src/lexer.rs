use std::borrow::Borrow;
use tabular_dfa::TabularDfa;
use super::grammar_parser::ast::{TokenDecl, SkipDecl};
use super::grammar_parser::regex::RegexError;
use super::{BiMap, Set};

pub use tokens::Tokens;

mod nfa;
mod nfa_builder;
mod dfa;
mod dfa_builder;
mod tabular_dfa;
mod powerset_cons;
mod util;
pub mod build;
pub mod tokens;

#[derive(Debug)]
pub struct Lexer {
  pub dfa: TabularDfa<TokenId>,
  pub char_intervals: Vec<u32>,
  pub skip: Set<TokenId>,
}

#[derive(PartialEq, Eq, PartialOrd, Ord, Hash, Clone, Copy, Debug)]
pub struct TokenId(u32);

#[derive(Default)]
pub struct TokenIdGen(u32);

#[derive(Debug)]
pub struct Token<'input> {
  pub kind: TokenId,
  pub text: &'input str,
  pub start: usize,
  pub end: usize,
}

impl TokenIdGen {
  pub fn gen(&mut self) -> TokenId {
    let i = self.0;
    self.0 += 1;
    TokenId(i)
  }

  pub fn from(start: u32) -> Self {
    Self(start + 1)
  }
}

#[derive(Debug)]
pub enum LexerError {
  RegexError(RegexError),
}

impl TokenId {
  pub fn id(&self) -> u32 {
    self.0
  }
}

impl Borrow<u32> for TokenId {
  fn borrow(&self) -> &u32 {
    &self.0
  }
}

impl Lexer {
  pub fn new(
    decls: &[TokenDecl],
    skips: &[SkipDecl]
  ) -> Result<(Self, BiMap<TokenId, String>), LexerError> {
    build::build(decls, skips)
  }

  pub fn lex<'lexer, 'input>(
    &'lexer self,
    input: &'input str
  ) -> Tokens<'lexer, 'input> {
    Tokens::new(self, input)
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use super::super::grammar_parser;
  use super::super::grammar_parser::ast::*;
  use insta::assert_debug_snapshot;

  #[test]
  fn skip() {
    let ast = grammar_parser::parse(r#"
%token INT /\d+/
%skip /[ \n]/
%skip /#[^\n]*/
    "#).unwrap();

    let decls = ast.iter()
      .filter_map(|decl| match &decl.1 {
        Decl::Token(decl) => Some(decl.clone()),
        _ => None,
      })
      .collect::<Vec<_>>();

    let skips = ast.iter()
      .filter_map(|decl| match &decl.1 {
        Decl::Skip(decl) => Some(decl.clone()),
        _ => None,
      })
      .collect::<Vec<_>>();

    let lexer = Lexer::new(&decls, &skips).unwrap().0;
    let tokens = lexer.lex(r"  123  456  # lorem ipsum
  0127401  #

 5768   ##dolorsitamet##
    #").collect::<Vec<_>>();

    assert_debug_snapshot!(tokens);
  }

  #[test]
  fn lex_expr() {
    let ast = grammar_parser::parse(r#"
%token PLUS "+"
%token MINUS "-"
%token MUL "*"
%token DIV "/"
%token LPAREN "("
%token RPAREN ")"
%token COMMA ","
%token NUMBER /\d+(\.\d*)?/
%token IDENT /[a-zA-Z][\w_]*/

%skip /[ \n]/
    "#).unwrap();

    let decls = ast.iter()
      .filter_map(|decl| match &decl.1 {
        Decl::Token(decl) => Some(decl.clone()),
        _ => None,
      })
      .collect::<Vec<_>>();

    let skips = ast.iter()
      .filter_map(|decl| match &decl.1 {
        Decl::Skip(decl) => Some(decl.clone()),
        _ => None,
      })
      .collect::<Vec<_>>();

    let lexer = Lexer::new(&decls, &skips).unwrap().0;
    let tokens = lexer.lex(r"(3.2 * 51 + Foo_1) / 20. -5  ,    ").collect::<Vec<_>>();

    assert_debug_snapshot!(tokens);
  }

  #[test]
  fn match_longest() {
    let ast = grammar_parser::parse(r#"
%token IN "in"
%token INTEGER "integer"
%token TEGE "tege"
    "#).unwrap();

    let decls = ast.iter()
      .filter_map(|decl| match &decl.1 {
        Decl::Token(decl) => Some(decl.clone()),
        _ => None,
      })
      .collect::<Vec<_>>();

    let skips = ast.iter()
      .filter_map(|decl| match &decl.1 {
        Decl::Skip(decl) => Some(decl.clone()),
        _ => None,
      })
      .collect::<Vec<_>>();

    let lexer = Lexer::new(&decls, &skips).unwrap().0;
    let tokens = lexer.lex(r"integeintegerinteg").collect::<Vec<_>>();

    assert_debug_snapshot!(tokens);
  }
}