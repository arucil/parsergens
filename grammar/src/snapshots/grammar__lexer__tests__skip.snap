---
source: grammar/src/lexer.rs
expression: tokens
---
[
    Ok(
        Token {
            kind: TokenId(
                0,
            ),
            text: "123",
            start: 2,
            end: 5,
        },
    ),
    Ok(
        Token {
            kind: TokenId(
                0,
            ),
            text: "456",
            start: 7,
            end: 10,
        },
    ),
    Ok(
        Token {
            kind: TokenId(
                0,
            ),
            text: "0127401",
            start: 28,
            end: 35,
        },
    ),
    Ok(
        Token {
            kind: TokenId(
                0,
            ),
            text: "5768",
            start: 41,
            end: 45,
        },
    ),
]
